Feature/gm communication import

guymaness2 opened the pull request:
### Description 

##### What does the change add or fix?

Add support for importing communications.

---------

### Release Notes 

##### What does the change add or fix in a succinct statement that will be read by clients?

* Added models and supporting logic to facilitate importing communications.

---------

### Requested By

##### Who reported, requested, or paid for the change?

Ada

---------

### Screenshots

##### Does this update or add options to the block UI?

no

---------

### Change Log

##### What files does it affect?

* Bulldozer.CSV/Bulldozer.CSV.csproj
* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/CSVInstance.cs
* Bulldozer.CSV/Maps/Communication.cs
* Bulldozer.CSV/Model/Communication.cs
* Bulldozer.CSV/Model/CommunicationRecipient.cs. 

---------

### Migrations/External Impacts

##### Is it a breaking change for other versions/clients?

no

nateh777 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                    CreatedByPersonAliasId = createdByPerson != null ? createdByPerson.PersonAliasId : ImportPersonAliasId,
+                    ForeignId = communicationCsv.CommunicationId.AsIntegerOrNull(),
+                    ForeignKey = $"{this.ImportInstanceFKPrefix}^{communicationCsv.CommunicationId}",
+                    CreatedDateTime = communicationCsv.DateCreated, 
```
```suggestion
                    CreatedDateTime = communicationCsv.CreatedDateTime, 
```

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                    Subject = communicationCsv.Subject,
+                    Message = communicationCsv.EmailMessage,
+                    SMSMessage = communicationCsv.SMSMessage,
+                    CommunicationType = ( CommunicationType ) ( int ) communicationCsv.CommunicationType,
```
Test with bogus (non empty) communicationtype string in csv.

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                    SMSMessage = communicationCsv.SMSMessage,
+                    CommunicationType = ( CommunicationType ) ( int ) communicationCsv.CommunicationType,
+                    SenderPersonAliasId = sender != null ? sender.PersonAliasId : ImportPersonAliasId,
+                    SendDateTime = communicationCsv.SentDateTime.GetValueOrDefault() < RockDateTime.Now ? communicationCsv.SentDateTime.GetValueOrDefault() : new DateTime( 1900, 1, 1 ), 
```
```suggestion
                    SendDateTime = communicationCsv.SentDateTime.GetValueOrDefault() < RockDateTime.Now ? communicationCsv.SentDateTime.GetValueOrDefault() : new DateTime( 1900, 1, 1 ), 
                    ReviewedDateTime = communicationCsv.SentDateTime.GetValueOrDefault() < RockDateTime.Now ? communicationCsv.SentDateTime.GetValueOrDefault() : new DateTime( 1900, 1, 1 ), 
```

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+
+            if ( communicationRecipientCsvsToProcess.Count() < this.CommunicationRecipientCsvList.Count() )
+            {
+                this.ReportProgress( 0, $"{this.CommunicationRecipientCsvList.Count() - communicationRecipientCsvsToProcess.Count()} CommunicationRecipients(s) from import already exist and will be skipped." );
```
```suggestion
                this.ReportProgress( 0, $"{this.CommunicationRecipientCsvList.Count() - communicationRecipientCsvsToProcess.Count()} CommunicationRecipients(s) have been skipped because they already exist or their communication id does not exist." );
```

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                .Where( c => c.ForeignKey != null && c.ForeignKey.StartsWith( this.ImportInstanceFKPrefix + "^" ) )
+                .ToList()
+                .ToDictionary( k => k.ForeignKey, v => v );
+
+            var communicationRecipientCsvsToProcess = this.CommunicationRecipientCsvList.Where( c => !importedCommunicationRecipients.ContainsKey( $"{this.ImportInstanceFKPrefix}^{c.CommunicationRecipientId}" ) && importedCommunications.ContainsKey( $"{this.ImportInstanceFKPrefix}^{c.CommunicationId}" ) );
```
```suggestion
```

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+
+            if ( missingCommunicationIds.Count > 0 )
+            {
+                LogException( $"CommunicationImport", $"The following invalid CommunicationId(s) in the communication recipient csv resulted in the following {missingCommunicationIds.Count} communication recipient(s) being skipped:\r\n{string.Join( ", ", missingPersonIds )}." );
```
```suggestion
                LogException( $"CommunicationImport", $"The following invalid CommunicationId(s) in the communication recipient csv resulted in the following {missingCommunicationIds.Count} communication recipient(s) being skipped:\r\n{string.Join( ", ", missingCommunicationIds )}." );
```

nateh777 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                }
+                else
+                {
+
```
```suggestion
```

guymaness2 commented on Bulldozer.CSV/Maps/Communication.cs:
```diff
@@ -0,0 +1,233 @@
+                    {
+                        CommunicationId = communication.Id,
+                        PersonAliasId = recipient.PersonAliasId,
+                        Status = ( CommunicationRecipientStatus ) ( int )  communicationRecipientCsv.RecipientStatus.GetValueOrDefault(),
```
```suggestion
                        Status = ( CommunicationRecipientStatus ) ( int ) communicationRecipientCsv.RecipientStatus.GetValueOrDefault(),
```

guymaness2 commented on Bulldozer.CSV/Maps/Financial.cs:
```diff
@@ -1169,7 +1169,20 @@ private int ImportFinancialAccounts()
                 {
-                    newAccount.CampusId = GetCampus( account.Campus.CampusId, this.ImportInstanceFKPrefix, UseExistingCampusIds, account.Campus.CampusName, true );
+                    var campusIdString = string.Empty;
+                    if ( account.Campus.CampusId.IsNotNullOrWhiteSpace() )
```
Can we move this logic into the GetCampus method?

guymaness2 commented on Bulldozer.CSV/Maps/Financial.cs:
```diff
@@ -1286,7 +1299,20 @@ private int ImportFinancialBatch()
                 {
-                    newBatch.CampusId = GetCampus( batch.Campus.CampusId, this.ImportInstanceFKPrefix, UseExistingCampusIds, batch.Campus.CampusName );
+                    var campusIdString = string.Empty;
+                    if ( batch.Campus.CampusId.IsNotNullOrWhiteSpace() )
```
Can we move this logic into the GetCampus method?

nateh777 commented on Bulldozer.CSV/Model/Communication.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public string CreatedByPersonId { get; set; }
+
+        public DateTime? DateCreated { get; set; }
```
```suggestion
        public DateTime? CreatedDateTime { get; set; }
```

guymaness2 commented:
@nateh777 all code is now working, know known issues

nateh777 commented on Bulldozer.CSV/CSVInstance.cs:
```diff
@@ -28,108 +28,6 @@ public class CSVInstance
     {
         #region Enums
 
-        /// <summary>
```
@guymaness2 this is quite the significant change, was this thoroughly tested? I think I understand that you are now just parsing directly to actual Rock Enums if you can, just hard coded strings are an interesting choice. Do we trim values and such to ensure that they get handled appropriately, even if they are valid? May have already been an issue casting to an enum, but I am curious if those are more forgiving on trimming values.

nateh777 commented on Bulldozer.CSV/Model/EntityAttributeCsv.cs:
```diff
@@ -14,7 +20,39 @@ public class EntityAttributeCsv
+            set
+            {
+                _attributeEntityType = value;
+                _isAttributeEntityType = Enum.TryParse( value.Trim().Replace( " ", string.Empty ), true, out _attributeEntityTypeEnum );
```
```suggestion
                _isValidAttributeEntityType = Enum.TryParse( value.Trim().Replace( " ", string.Empty ), true, out _attributeEntityTypeEnum );
```

nateh777 commented on Bulldozer.CSV/Model/EntityAttributeCsv.cs:
```diff
@@ -14,7 +20,39 @@ public class EntityAttributeCsv
+            }
+        }
+
+        public bool IsValidRecordStatus
+        {
+            get
+            {
+                return _isAttributeEntityType;
+            }
+        }
```
```suggestion
        public bool IsValidAttributeEntityType
        {
            get
            {
                return _isValidAttributeEntityType;
            }
        }
```

nateh777 commented on Bulldozer.CSV/Model/ScheduleCsv.cs:
```diff
@@ -5,13 +5,49 @@ namespace Bulldozer.Model
     {
+        private string _meetingDay = string.Empty;
+        private DayOfWeek _meetingDayEnum = DayOfWeek.Sunday;
+        private bool _isValidmeetingDay = false;
```
```suggestion
        private bool _isValidMeetingDay = false;
```

nateh777 commented on Bulldozer.CSV/Model/ScheduleCsv.cs:
```diff
@@ -5,13 +5,49 @@ namespace Bulldozer.Model
+            set
+            {
+                _meetingDay = value;
+                Enum.TryParse( value.Trim().Replace( " ", string.Empty ), true, out _meetingDayEnum );
```
```suggestion
                _isValidMeetingDay = Enum.TryParse( value.Trim().Replace( " ", string.Empty ), true, out _meetingDayEnum );
```

nateh777 commented on Bulldozer.CSV/Model/ScheduleCsv.cs:
```diff
@@ -5,13 +5,49 @@ namespace Bulldozer.Model
+            }
+        }
+
+        public bool IsValidmeetingDay
+        {
+            get
+            {
+                return _isValidmeetingDay;
```
```suggestion
        public bool IsValidMeetingDay
        {
            get
            {
                return _isValidMeetingDay;
```

nateh777 commented on Bulldozer.CSV/Maps/Business.cs:
```diff
@@ -200,15 +200,17 @@ public int BulkBusinessImport( RockContext rockContext, List<BusinessCsv> busine
                     Email = businessCsv.Email,
                     IsEmailActive = businessCsv.IsEmailActive.HasValue ? businessCsv.IsEmailActive.Value : true,
-                    EmailPreference = businessCsv.EmailPreference.HasValue ? businessCsv.EmailPreference.Value : EmailPreference.EmailAllowed,
+                    EmailPreference = businessCsv.EmailPreferenceEnum.HasValue ? businessCsv.EmailPreferenceEnum.Value : EmailPreference.EmailAllowed,
```
```suggestion
                    EmailPreference = businessCsv.EmailPreferenceEnum,
```

nateh777 commented on Bulldozer.CSV/Maps/Business.cs:
```diff
@@ -221,6 +223,19 @@ public int BulkBusinessImport( RockContext rockContext, List<BusinessCsv> busine
+
+                if ( !validRecordStatus )
+                {
+                    errors += string.Format( "{0},{1},\"{2}\"\r\n", DateTime.Now.ToString(), "Business", string.Format( "Unexpected RecordStatus ({0}) encountered for BusinessId \"{1}\". Record Status defaulted to \"{2}\".", businessCsv.RecordStatus, businessCsv.Id, businessCsv.RecordStatusEnum.ToString() ) );
```
```suggestion
                    errors += string.Format( "{0},{1},\"{2}\"\r\n", DateTime.Now.ToString(), "Business", string.Format( "Unexpected RecordStatus ({0}) encountered for BusinessId \"{1}\". Record Status defaulted to \"Inactive\".", businessCsv.RecordStatus, businessCsv.Id ) );
```

nateh777 commented on Bulldozer.CSV/Maps/Person.cs:
```diff
@@ -535,14 +539,24 @@ private int ImportPersonAddresses()
                 {
-                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Unexpected Address Type ({addressCsv.PersonAddressCsv.AddressType}) encountered for Person \"{addressCsv.PersonAddressCsv.PersonId}\". Person Address was skipped.\r\n";
+                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Unexpected Address Type ({addressCsv.PersonAddressCsv.AddressTypeEnum}) encountered for Person \"{addressCsv.PersonAddressCsv.PersonId}\". Person Address was skipped.\r\n";
+                }
+                if ( string.IsNullOrEmpty( addressCsv.PersonAddressCsv.Street1 ) )
+                {
+                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Blank Street Address for PersonId {addressCsv.PersonAddressCsv.PersonId}, Address Type {addressCsv.PersonAddressCsv.AddressTypeEnum}. Person Address was skipped.\r\n";
+                    continue;
+                }
+                if ( addressCsv.Family == null )
+                {
+                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Family for PersonId {addressCsv.PersonAddressCsv.PersonId} not found. Person Address was skipped.\r\n";
+                    continue;
```
```suggestion
```

nateh777 commented on Bulldozer.CSV/Maps/Person.cs:
```diff
@@ -501,32 +516,21 @@ private int ImportPersonAddresses()
-                }
-
-                var groupLocationTypeValueId = GetGroupLocationTypeDVId( addressCsv.PersonAddressCsv.AddressType );
+                var groupLocationTypeValueId = GetGroupLocationTypeDVId( addressCsv.PersonAddressCsv.AddressTypeEnum.Value );
```
```suggestion
if ( string.IsNullOrEmpty( addressCsv.PersonAddressCsv.Street1 ) )
                {
                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Blank Street Address for PersonId {addressCsv.PersonAddressCsv.PersonId}, Address Type {addressCsv.PersonAddressCsv.AddressTypeEnum}. Person Address was skipped.\r\n";
                    continue;
                }
                if ( addressCsv.Family == null )
                {
                    familyAddressErrors += $"{DateTime.Now}, PersonAddress, Family for PersonId {addressCsv.PersonAddressCsv.PersonId} not found. Person Address was skipped.\r\n";
                    continue;
                }
                
                var groupLocationTypeValueId = GetGroupLocationTypeDVId( addressCsv.PersonAddressCsv.AddressTypeEnum.Value );
```

nateh777 commented on Bulldozer.CSV/Maps/Group.cs:
```diff
@@ -1121,14 +1139,19 @@ public int BulkGroupMemberImport( RockContext rockContext, List<GroupMemberCsv>
                 }
+                if ( !groupMemberCsv.IsValidGroupMemberStatus )
+                {
+                    groupMemberErrors += $"{DateTime.Now}, GroupMember, Unexpected GroupMemberStatus ({groupMemberCsv.GroupMemberStatus}) encountered encountered for PersonId \"{groupMemberCsv.PersonId}\" in GroupId \"{groupMemberCsv.GroupId}\". Group Member Status was defaulted to {groupMemberCsv.GroupMemberStatusEnum}.\r\n";
```
```suggestion
                    groupMemberErrors += $"{DateTime.Now}, GroupMember, Unexpected GroupMemberStatus ({groupMemberCsv.GroupMemberStatus}) encountered for PersonId \"{groupMemberCsv.PersonId}\" in GroupId \"{groupMemberCsv.GroupId}\". Group Member Status was defaulted to {groupMemberCsv.GroupMemberStatusEnum}.\r\n";
```

guymaness2 commented on Bulldozer.CSV/Maps/GroupMemberHistorical.cs:
```diff
@@ -163,7 +163,7 @@ public int BulkInsertGroupMembersFromImport( RockContext rockContext, List<Group
                     GroupTypeId = group.GroupTypeId,
                     RoleName = groupMemberHistoricalCsv.Role,
-                    GroupMemberStatus = groupMemberHistoricalCsv.GroupMemberStatusCurrent.GetValueOrDefault(),
+                    GroupMemberStatus = groupMemberHistoricalCsv.GroupMemberStatusCurrentEnum.GetValueOrDefault(),
```
```suggestion
                    GroupMemberStatus = groupMemberHistoricalCsv.GroupMemberStatusCurrentEnum.Value,
```

guymaness2 commented on Bulldozer.CSV/Maps/GroupMemberHistorical.cs:
```diff
@@ -172,6 +172,11 @@ public int BulkInsertGroupMembersFromImport( RockContext rockContext, List<Group
 
+                if ( !groupMemberHistoricalCsv.IsValidCurrentGroupMemberStatus )
+                {
+                    groupMemberErrors += $"{DateTime.Now}, GroupMember, Invalid GroupMemberStatusHistorical value ({groupMemberHistoricalCsv.GroupMemberStatusCurrent}) provided for PersonId {groupMemberHistoricalCsv.PersonId}, GroupId {groupMemberHistoricalCsv.GroupId}. GroupMemberStatus was defaulted to \"{groupMemberHistoricalCsv.GroupMemberStatusCurrentEnum.Value}\".\r\n";
```
```suggestion
                    groupMemberErrors += $"{DateTime.Now}, GroupMember, Invalid GroupMemberStatusCurrent value ({groupMemberHistoricalCsv.GroupMemberStatusCurrent}) provided for PersonId {groupMemberHistoricalCsv.PersonId}, GroupId {groupMemberHistoricalCsv.GroupId}. GroupMemberStatus was defaulted to \"{groupMemberHistoricalCsv.GroupMemberStatusCurrentEnum.Value}\".\r\n";
```

nateh777 commented on Bulldozer.CSV/Maps/Schedule.cs:
```diff
@@ -69,6 +68,10 @@ public int ImportSchedules()
                     ForeignId = scheduleCsv.Id.AsIntegerOrNull(),
                     IsActive = scheduleCsv.IsActive
                 };
+                if ( scheduleCsv.IsValidmeetingDay )
```
```suggestion
                if ( scheduleCsv.IsValidMeetingDay )
```

nateh777 commented on Bulldozer.CSV/Maps/Person.cs:
```diff
@@ -273,6 +275,19 @@ public int BulkPersonImport( RockContext rockContext, List<PersonCsv> personCsvs
+                }
+                if ( !personCsv.IsValidEmailPreference )
+                {
+                    errors += string.Format( "{0},{1},\"{2}\"\r\n", DateTime.Now.ToString(), "Person", string.Format( "Unexpected Email Preference ({0}) encountered for PersonId \"{1}\". Email Preference defaulted to \"{2}\".", personCsv.EmailPreference, personCsv.Id, personCsv.EmailPreferenceEnum.ToString() ) );
```
```suggestion
                    errors += string.Format( "{0},{1},\"{2}\"\r\n", DateTime.Now.ToString(), "Person", string.Format( "Unexpected Email Preference ({0}) encountered for PersonId \"{1}\". Email Preference defaulted to \"Inactive\".", personCsv.EmailPreference, personCsv.Id) );
```

nateh777 commented on Bulldozer.CSV/Maps/Business.cs:
```diff
@@ -394,6 +398,17 @@ private int ImportBusinessAddresses()
                 {
                     familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Unexpected Address Type ({addressCsv.BusinessAddressCsv.AddressType}) encountered for BusinessId \"{addressCsv.BusinessAddressCsv.BusinessId}\". Business Address was skipped.\r\n";
                 }
+
+                if ( string.IsNullOrEmpty( addressCsv.BusinessAddressCsv.Street1 ) )
+                {
+                    familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Blank Street Address for BusinessId {addressCsv.BusinessAddressCsv.BusinessId}, Address Type {addressCsv.BusinessAddressCsv.AddressType}. Business Address was skipped.\r\n";
+                    continue;
+                }
+                if ( addressCsv.Family == null )
+                {
+                    familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Family for BusinessId {addressCsv.BusinessAddressCsv.BusinessId} not found. Business Address was skipped.\r\n";
+                    continue;
+                }
```
```suggestion
```

nateh777 commented on Bulldozer.CSV/Maps/Business.cs:
```diff
@@ -351,32 +366,21 @@ private int ImportBusinessAddresses()
-                    familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Family for BusinessId {addressCsv.BusinessAddressCsv.BusinessId} not found. Business Address was skipped.\r\n";
-                    continue;
-                }
+                var groupLocationTypeValueId = GetGroupLocationTypeDVId( addressCsv.BusinessAddressCsv.AddressTypeEnum.Value );
```
```suggestion

                if ( string.IsNullOrEmpty( addressCsv.BusinessAddressCsv.Street1 ) )
                {
                    familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Blank Street Address for BusinessId {addressCsv.BusinessAddressCsv.BusinessId}, Address Type {addressCsv.BusinessAddressCsv.AddressType}. Business Address was skipped.\r\n";
                    continue;
                }
                if ( addressCsv.Family == null )
                {
                    familyAddressErrors += $"{DateTime.Now}, BusinessAddress, Family for BusinessId {addressCsv.BusinessAddressCsv.BusinessId} not found. Business Address was skipped.\r\n";
                    continue;
                }
                
                var groupLocationTypeValueId = GetGroupLocationTypeDVId( addressCsv.BusinessAddressCsv.AddressTypeEnum.Value );
```

guymaness2 commented:
@nateh777 
The following changes have been made to the wiki on 7/3/2024 to support the changes from this pull request:
* Added communication.csv documentation to content and navigation.
* Added communicaiton-recipient.csv documentation to content and navigation.
* Updated grouptype.csv content to include additional fields now supported.
* Updated csv content for various sections to describe defaulting behavior of some enum based fields.
  
 The following changes were also made to the wiki based on changes from previous pull requests:
* Renamed batch.csv section to financial-batch.csv and updated content to match updated model.
* Renamed account.csv section to financial-account.csv. and updated content to match updated model.
* Added groupmember-historical.csv documentation to content.
* Update navigation to reflect changes above.

nateh777 commented on Bulldozer.CSV/Maps/Business.cs:
```diff
@@ -351,11 +366,12 @@ private int ImportBusinessAddresses()
 
             foreach ( var addressCsv in addressCsvObjectsToProcess )
             {
+
```
```suggestion
```

nateh777 commented on Bulldozer.CSV/Maps/Person.cs:
```diff
@@ -501,32 +516,32 @@ private int ImportPersonAddresses()
             foreach ( var addressCsv in addressCsvObjectsToProcess )
             {
-                if ( string.IsNullOrEmpty( addressCsv.PersonAddressCsv.Street1 ) )
+if ( string.IsNullOrEmpty( addressCsv.PersonAddressCsv.Street1 ) )
```
```suggestion
                if ( string.IsNullOrEmpty( addressCsv.PersonAddressCsv.Street1 ) )
```
